README.rst
.. image:: https://travis-ci.org/timevortexproject/timevortex.svg?branch=develop
    :target: https://travis-ci.org/timevortexproject/timevortex

.. image:: https://coveralls.io/repos/github/timevortexproject/timevortex/badge.svg?branch=develop
    :target: https://coveralls.io/github/timevortexproject/timevortex?branch=develop 

.. image:: https://codeclimate.com/github/timevortexproject/timevortex/badges/gpa.svg
    :target: https://codeclimate.com/github/timevortexproject/timevortex
    :alt: Code Climate

Setup
=====

* mkdir timevortex_v2
* mkvirtualenv timevortex_v2 --python=/usr/bin/python3
* sudo apt-get install python3 python3-dev

Install
=======

Prod
----

* sudo apt-get install python3 python3-dev sqlite3
* pip install django requests python-dateutil pytz django-admin-tools psutil pyserial
* mkdir /var/log/timevortex

Dev
---
    
* sudo apt-get install socat git-flow
* pip install flake8 pylint django_nose behave_django pylint-django coverage prospector nose-exclude
* [BLOCKED] pip install clonedigger django-fluent-dashboard django-admin-tools-stats
* mkdir /tmp/timevortex

Create application
==================

* Create Django apps:
    * django-admin startproject timevortex
    * python manage.py startapp weather
    * python manage.py startapp hardware
    * python manage.py startapp energy
    * python manage.py startapp stubs
    * python manage.py makemigrations weather
    * python manage.py migrate
    * python manage.py collectstatic
    * python manage.py createsuperuser --settings=timevortex.settings.local
        * timevortex / timevortex_Admin
* Manage project
    * python manage.py timevortex --commit "Your commit message" --settings=timevortex.settings.local


Snap build
==========

* sudo apt-get install software-properties-common
* sudo add-apt-repository ppa:snappy-dev/tools
* sudo apt-get update
* sudo apt-get install snapcraft
* snapcraft init (create a yml file)
* snapcraft stage (retrieve source and install dependencies)
* snapcraft snap (build snap)
* snapcraft (build final snap package)
* sudo apt-get install snapd
* sudo snap install timevortex_2.0.1_amd64.snap --devmode --force-dangerous
* sudo snap try prime --devmode
* cp /vagrant/snapcraft.yaml . && cp /vagrant/timevortex-prepare . && snapcraft clean && snapcraft && sudo snap try prime --devmode

TimeVortex todo
===============

* Prod :
    * Put snap folder in timevortex folder
    * Install ubuntu 16.04 on raspberry
    * Test daily-report
    * Test backup in USB key
    * Test currentcost live
    * Remove develop branch in snapcraft.yml
    * Migrate data to new currentcost
    * Setup raspberry pi installation at home

TimeVortex backlog
==================

* [BLOCKED]
    * [BLOCKED] django-admin-tools-stats is not python 3 compatible

* Documentation
    * Update documentation for this project
    * Generate documentation from code
    * Generate document from .rst file

* CSV crawler
    * Finance     (CSV)
    * Garmin data (CSV)
* API
    * Fitbit      (API) => http://python-fitbit.readthedocs.org/en/latest
    * Withing     (API) => https://github.com/maximebf/python-withings
    * Jawbone     (API) => http://kiefer.readthedocs.org/en/latest
    * Runkeeper   (API) 

* Hardware
    * Retrieve CPU data
    * Retrieve Memory data
    * Retrieve HDD data

Timevortex idea
---------------

* Global
    * Split command into several celery job
    * Refactor metear script to include celery script to crawl the page and collect data. Then create a json element with "siteID", "variableID", "values" = [(value, date, dsttimezone, nondsttimezone)] that send 48 values to TSL with only one message. Store it into a CSV file and into DB. This should improve time to retrieve data and improve performance. Wait to install performance tracking before improving this script.
    * Use petl to manipulate data.
    * Create à fonction that take into paramètres à matrix of variable and date and insert easily data in DB and send message over the network. 
        |       |  var1  | var2   |
        | time1 | value1 | value2 |
        | time2 | value3 | value4 |
    * Use vagrant David config
* Technology
    * Microservice with python and django, microservice django
    * Docker, host docker image

* Idea:
    * Force user to register with gmail account. Use this gmail account to send email for daily report. Email account should be sender and receiver
    * Use google SSO to register a user and retrieve information about him
    * Define a flow to register and configure an account into timevortex paltform and create functional tests based on this flow
    * user should select a city where he actually leaves. City selection propose lang and unit that user want to use. By default lang is browser language, unit is metrics system

* UI:
    * Chart with aggregation to 3 months (temperature, kWh, split hp/hc)
    * Chart with aggreggation to 1 month (temperature, kWh, split hp/hc)
    * Chart per week
    * Chart per day

* Use cases:
    * CRUD user into platform
    * CRUD site information
    * CRUD sensor (adding a sensor create several variables. User is free to add or not a variable)
    * CRUD variables
    * Link variable to a room or to a site
    * Choose representation (timeline or site plan with top view and all variable all around)
    * Optional : A variable is the combination of several variable

* Adding a sensor :
    * Choice by brand and model
    * Define parameter for the sensor
    * List of generated variable

* Use django as plugin provider to reduce time to create a plugin and simplify deployment
* Each django app should create a django command that use RBMQ and define a REST API to retrieve data
* Create a Django app per thematics like weather, electricity, finance, health
* Create a Djnago stubs app for stubs that could be activated by settings

Consulting
-----------------
    * Être quelqu'un qui aide les gens à y voir plus clair dans leurs données
    * Proposer un système automatique de collecte et d'analyse de différents types de données
    * Proposer des simulations pour améliorer certains points
    * Faire la liste des points qui pourraient être améliorer
    * Proposer un blog autour des améliorations possibles
    * Proposer des jeux afin de se connaitre mieux
    * Ces jeux doivent être ouvert à tous sous excell par exemple
    * Apprendre en s'amusant pour attirer les gens
    * Ensuite rendre une solution packages pour automatiser tous ça
    * Ouvrir un blog sur le quantified self est essayé de le démocratiser





CHANGELOG.rst
TimeVortex (2.1.2) stable; urgency=low

* Add snap folder into TV v2 project
* Add snap folder into TV v2 project
* Load initial data without erasing user modification
* Fix energy model problem
* Pass log error if no currentcost settings to 60 seconds, Log into system site_id
* Migrate command parameter of currentcost to DB settings
* Fix lint problem
* Fix migration problem
* Improve daily_report to read settings from DB, send report every day at 4:00 AM
* Migrate abstract command to infinite loop and update job to run new abstract command architecture
* Fix lint problem
* Move Settings weather value DateField to Charfield
* Move BACKUP_TARGET_FOLDER in DB + add initial_data + fix funtional test for backup command
* Move SETTINGS_METEAR_START_DATE in DB + add initial_data + fix funtional test + fix bad url problem in metear command
* Remove useless log from timevortex model, Add todo for backup and metear command
* Move METEAR command to infinite loop
* Try to export static files
* Try to export static files
* Try to export static files
* Add initial_data and test to start server on deployment
* Add initial_data and test to start server on deployment
* Add energy and weather command
* Fix lint problemp
* Try to export energy folder
* Try to export energy folder
* Remove stubs app form production, only launch stubs url in testing mode
* Remove stubs app form production, only launch stubs url in testing mode
* Reimport stubs in base.py
* Change DB settings to try to migrate
* Fix lint problem
* Fix base.py and local.py settings to add unit test only on local and not in production
* Add dependency dev in setup.py for test
* Add djnago_nose for test
* Exclude stubs package
* Exclude stubs package
* Modify production settings to target snap common path
* Modify production settings to target snap common path
* Modify python file to use python3
* Start release v2.1.2

Pierre Leray <pierreleray64@gmail.com>  2016-09-20 12:52:29

TimeVortex (2.1.1) stable; urgency=low

* Finish release v2.1.1 on 2016-09-20 12:51:40
* Modify setup.py to export manage.py
* Start release v2.1.1

Pierre Leray <pierreleray64@gmail.com>  2016-09-15 11:38:49

TimeVortex (2.1.0) stable; urgency=low

* Finish release v2.1.0 on 2016-09-15 11:38:49
* Add setup.py
* Start release v2.1.0

Pierre Leray <pierreleray64@gmail.com>  2016-09-09 13:28:56

TimeVortex (2.0.1) stable; urgency=low

* Finish release v2.0.1 on 2016-09-09 13:28:56
* Fix pep8 error
* Fix pep8 error
* Fix behave error
* Fix pylint error
* Modify travis.yml to create folder
* Add local.py for testing, modify gitignore
* Fix Django version in requirements
* Modifiy requirements to pass travis test, update test to fix path problem
* Fix requirements in travis.yml
* Fix requirements in travis.yml
* Fix lint, test and directory on local branch
* Add Snap creation command on readme, update travis.yml to read settigns from local
* Start release v2.0.1

Pierre Leray <pierreleray64@gmail.com>  2016-07-18 00:36:18

TimeVortex (2.0.0) stable; urgency=low

* Finish release v2.0.0 on 2016-07-18 00:36:18
* Migrate TimeVortex application to Django
* Create timevortex app for utils and timeserieslogger
* Create energy app for currentcost command
* Create weather app for METEAR command
* Migrate pavement utils to timevortex command

Pierre Leray <pierreleray64@gmail.com>  2016-07-17 23:18:43




temp.py
PACKAGE = "timevortex"

options(
    sphinx=Bunch(
        builddir="build",
        sourcedir="source"
    ),

    bin=PACKAGE,

    test_package="tests",

    clone_file="reports/clone/index.html",

    stats_file="reports/stats/index.html",

    cover_folder="reports/cover",

    pylint_file=".pylintrc",

    files="*.py %s/*.py bin/%s features/steps/*.py" % (PACKAGE, PACKAGE),
)


def create_new_release(new_version):
    """Create new release adding a message on changelog, updating version
    in setup.py and conf.py.
    """
    regex = '16s/^.*$/VERSION = "%s"/g' % new_version
    # Update setup.py version
    sh("sed -i '%s' setup.py" % regex)
    # Update conf.py version
    sh("sed -i '%s' docs/source/conf.py" % regex)
    # Add new release message in changelog
    message_starting = "Starting release %s" % new_version
    command = "dch --newversion %s %s" % (new_version, message_starting)
    sh(command)


def get_version(release_type):
    """This method return actual version of package and the next according
    to release_type
    """
    version = release_type
    temp_file = "version.package.txt"
    sh("python setup.py -V > %s" % temp_file)
    with open(temp_file) as filename:
        actual_version = filename.readline().replace("\n", "")
        actual_version = actual_version.replace("\r", "")
        filename.close()
    sh("rm %s" % temp_file)
    array_actual_version = actual_version.split(".")

    if len(array_actual_version) != 3:
        array_actual_version = ["0", "0", "0"]

    if version == "micro":
        new_version = "%s.%s.%d" % (
            array_actual_version[0],
            array_actual_version[1],
            int(array_actual_version[2]) + 1)
    elif version == "minor":
        new_version = "%s.%d.%d" % (
            array_actual_version[0],
            int(array_actual_version[1]) + 1,
            0)
    elif version == "major":
        new_version = "%d.%d.%d" % (int(array_actual_version[0]) + 1, 0, 0)
    else:
        sh("exit(-1)")

    return actual_version, new_version


@task
@consume_args
def release_stable(args):
    """Publish stable release:
        * push package on pip,
        * push code on master git branch with according tag
        * push code on master bazaar branch with according tag
        * test master pip package
        * indicate build status and pip package in README
        * return on develop branch
        * generate debian package
        * test debian package
        * switch to next release
    """
    if len(args) != 1:
        print("Missing next release type argument. Please provide a type.")
        return(0)
    release_type = args[0]
    actual_version, next_version = get_version(release_type)
    message_ending = "Finish release %s" % actual_version
    # Upload package on Pypi
    upload()
    # Create release branch and close it
    sh("git flow release start %s" % actual_version)
    sh("git flow release finish -m '%s' %s" % (
        message_ending, actual_version))
    # Push master branch to github and return on develop
    sh("git push origin master")
    sh("git co develop")
    # Commit code on bazaar and push it
    sh("bzr push '%s'" % (BZR_REPO))
    # Switch to next release
    create_new_release(next_version)

TimeVortex (2.0.0) stable; urgency=low

* Migrate TimeVortex application to Django
* Create timevortex app for utils and timeserieslogger
* Create energy app for currentcost command
* Create weather app for METEAR command
* Migrate pavement utils to timevortex command

Pierre Leray <pierreleray64@gmail.com>  2016-07-17 23:18:43




@task
@consume_args
def release(args):
    """Prepare next release taking into account message in parameter

    """
    # Variables
    version = args[0]
    commit = args[1]
    day = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    # Create new version number from previous tag
    filename = "tmp.txt"
    cmd = "git describe --tags `git rev-list --tags --max-count=1` > %s"
    sh(cmd % filename)
    fil = open(filename)
    tag_version = fil.readlines()
    fil.close()
    sh("rm %s" % filename)
    tag_version = tag_version[0].replace("\n", "")
    old_version = tag_version.replace("v", "").split(".")

    if len(old_version) != 3:
        old_version = ["0", "0", "0"]

    if version == "micro":
        new_version = "v%s.%s.%d" % (
            old_version[0],
            old_version[1],
            int(old_version[2]) + 1)
    elif version == "minor":
        new_version = "v%s.%d.%d" % (
            old_version[0],
            int(old_version[1]) + 1,
            0)
    elif version == "major":
        new_version = "v%d.%d.%d" % (int(old_version[0]) + 1, 0, 0)
    else:
        sh("exit(-1)")
    # Update changelog
    changelog = "timevortex.%s (%s) stable; urgency=low\n\n" % (
        PACKAGE, new_version)
    changelog += "* %s\n\n" % commit
    changelog += "Pierre Leray <pierreleray64@gmail.com>  %s\n" % day
    cmd = "echo '%s' | cat - CHANGELOG.rst > /tmp/out"
    cmd += " && mv /tmp/out CHANGELOG.rst"
    sh(cmd % changelog)
    # Update setup.py version
    regex = "9s/^.*$/VERSION = '%s'/g" % new_version[1:]
    sh("sed -i '%s' setup.py" % regex)
    # Update conf.py version
    regex = "67s/^.*$/version = '%s'/g" % ".".join(
        new_version[1:].split(".")[:2])
    sh("sed -i '%s' docs/source/conf.py" % regex)
    regex = "69s/^.*$/release = '%s'/g" % new_version[1:]
    sh("sed -i '%s' docs/source/conf.py" % regex)
    # Deploy new release on PYPi
    upload()
    # Commit change
    sh("git add . && git ci -a -m '%s' && git push origin develop" % commit)
    # Create release branch and close it
    sh("git flow release start %s" % new_version)
    sh("git flow release finish -m '%s' %s" % (commit, new_version))
    # Push master branch to github and return on develop
    sh("git push origin master")
    sh("git co develop")
